### PR TITLE
Base support for field type INTERVAL

### PR DESCRIPTION
**Pull request by ImeevMA**
This patch-set introduces base support of field type INTERVAL to BOX.

This is the first part of addition of field type INTERVAL.

### CODE REVIEW

File: src/lib/core/interval.h [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	/** See the description of DT_EXCESS in c-dt. */
+	INTERVAL_ADJUST_EXCESS,
+	/** See the description of DT_LIMIT in c-dt. */
+	INTERVAL_ADJUST_LIMIT,
+	/** See the description of DT_SNAP in c-dt. */
+	INTERVAL_ADJUST_SNAP,
+};
+
+/** A structure that describes a date and time interval. */
+struct interval {
```
Comment (tsafin) [inline src/lib/core/interval.h:41]: Could you please rebase your incomplete interval implementation upon full C reimplementation from #6923?

File: src/lib/core/interval.h [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	/** Number of weeks. */
+	int32_t week;
+	/** Number of days. */
+	int32_t day;
+	/** Number of hours. */
+	int32_t hour;
+	/** Number of minutes. */
+	int32_t min;
+	/** Number of seconds. */
+	int32_t sec;
```
Comment (tsafin) [inline src/lib/core/interval.h:55]: Here is the explanation why it's `double` in #6923 :
 - let assume we are subtracting 2 datetime with fields `epoch` (which is `double`) and `nsec` (which is `int`);
 - what type of result should be for '+' or '-' between 2 `epoch`s?

`int` is not sufficient enough to handle those values we could save there, only `int64_t` (which is boxed and produces performance impact due to GC) or `double` (which is unboxed, and native for Lua).

That's why we have to use same type as in `epoch`. 

OTOH, we could try to decompress full seconds difference into `year`, '`month`, 'day`, etc. But in this case __every__ arithmetic operation would require full cycle of normalization (including leap seconds, leap years, timezone differences, and what not) which is absolutely unnecessary here, and might be entirely avoided if simply save difference with `epoch` type...

**Review by tsafin — CHANGES_REQUESTED** [review]

File: src/lib/core/interval.h (hunk 2 above)
Comment (ImeevMA) [reply src/lib/core/interval.h:55]: We shouldn't subtract one epoch from another directly.

**Review by Gerold103 — CHANGES_REQUESTED** [review]
Thanks for the patch! I suppose the msgpack and SQL parts of it are valid mostly. But the problem is that we indeed already have intervals implementation for C and Lua in Timur's PR. With much richer set of operations, many more tests. As I was told, that PR is not discarded. It is going to be merged soon. That raises a concern why is it needed to implement the intervals second time from scratch and without operations at all? You can base your changes on top of Timur's. Just fetch his changes and apply yours in next commits. Yes, your PR will contain his commits too until they are in master, but not a big deal, we can just ignore them in the review.

I suggest to put aside the arguing like how to store epoch and shit. SQL shouldn't care about that anyway. Like it, for instance, doesn't care about content of `decimal_t` or `struct datetime`. You will have `struct interval` and some functions on it doing arithmetics and pack/unpack, internals are not important since they are already done in Timur's PR anyway.

The alternative ways like actually having 2 interval implementations in master in parallel, or one for SQL and other for Lua are not even worth considering. Much more complicated support, tons of duplicate code and tests, and inconsistency in public code about `datetime` module and results of `SELECT`. There of course should be only one.

**Comment by ImeevMA** [comment]
Thanks for the review! The problem with PR https://github.com/tarantool/tarantool/pull/6923 is not in tests or anything like that. First, we decided to do a quick refactoring "from Lua to C", on top of which we would have to write a new implementation of INTERVAL. However, during the review, new patches appeared, which now added new functionality. By the time these patches appeared, I had already written my implementation of INTERVAL. I planned to rebase this PR earlier since we didn't have a conflict, but that's not the case now. At the moment I'm still deciding what to do next. Also there won't be two implementations anyway. Either I will use Timur's intervals, or I will replace them.

**Pull request closed without merging**